Задачи по Spring Data JPA

Sindicat007 opened the pull request.

Sindicat007 committed:
Задачи по Spring Data JPA

Signed-off-by: Sindicat007 <sindicat007@ya.ru>

RinatBeybutov commented on Jpa/src/main/java/ru/maxima/jpa/controller/AppController.java:
```diff
@@ -0,0 +1,46 @@
+@RequestMapping("/")
+public class AppController {
+    public AppController(BookService bookService) {
+        this.bookService = bookService;
```
давай заменим на ломбок и @RequiredArgsConstructor

RinatBeybutov commented on Jpa/src/main/java/ru/maxima/jpa/controller/AppController.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.Optional;
+
+@RestController
+@RequestMapping("/")
```
общее (а именно books) выносим в RequestMapping над классом

RinatBeybutov commented on Jpa/src/main/java/ru/maxima/jpa/controller/AppController.java:
```diff
@@ -0,0 +1,46 @@
+    @GetMapping("/books")
+    public ResponseEntity<String> findBooks() {
+        List<BookDto> books = bookService.getAllBooks();
+        return ResponseEntity.ok(books.stream().map(BookDto::toString).toList().toString());
```
многовложенные вызовы - нечитабельный код, в несколько строк разнеси, каждая операция стрима с отдельной строки. А вообще у тебя в контроллере не должно быть никакой логики по преобразованию, из сервиса приходит ответ, который ты запихиваешь в ResponseEntity

RinatBeybutov commented on Jpa/src/main/java/ru/maxima/jpa/controller/AppController.java:
```diff
@@ -0,0 +1,46 @@
+
+    @GetMapping("/books/get/{id}")
+    public ResponseEntity<String> findBookById(@PathVariable("id") Long id) {
+        Optional<BookDto> book = bookService.getUserById(id);
```
а если empty придет, что делать?

RinatBeybutov commented on Jpa/src/main/java/ru/maxima/jpa/repository/BookRepository.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.stereotype.Repository;
+import ru.maxima.jpa.entity.Book;
+
+@Repository
```
у наследников JpaRepository не обязательно эту аннотацию ставить

Sindicat007 committed:
- Заменил инициализацию в контроллере на RequiredArgsConstructor
- Изменил путь requestMapping на /books
- Оптимизировал код в методе findBooks контроллера AppController
- Добавил обработку для несуществующих книг в методе findBookById контроллера AppController
- Убрал в репозитории аннотацию @Repository

Signed-off-by: Sindicat007 <sindicat007@ya.ru>

RinatBeybutov commented on Jpa/src/main/java/ru/maxima/jpa/controller/AppController.java:
```diff
@@ -1,45 +1,43 @@
     private final BookService bookService;
 
-    @GetMapping("/books")
+    @GetMapping("/getAllBooks")
```
зачем добавил getAllBooks?
убери

RinatBeybutov commented on Jpa/src/main/java/ru/maxima/jpa/controller/AppController.java:
```diff
@@ -1,45 +1,43 @@
     }
 
-    @GetMapping("/books/add")
+    @GetMapping("/add")
```
PostMapping, убери add

RinatBeybutov commented on Jpa/src/main/java/ru/maxima/jpa/controller/AppController.java:
```diff
@@ -1,45 +1,43 @@
     }
 
-    @GetMapping("/books/delete/{id}")
+    @GetMapping("/delete/{id}")
```
DeleteMapping убери delete

RinatBeybutov commented on Jpa/src/main/java/ru/maxima/jpa/controller/AppController.java:
```diff
@@ -1,45 +1,43 @@
     }
 
-    @GetMapping("/books/get/{id}")
+    @GetMapping("/get/{id}")
```
убери get

RinatBeybutov commented on Jpa/src/main/java/ru/maxima/jpa/controller/AppController.java:
```diff
@@ -0,0 +1,44 @@
+import ru.maxima.jpa.service.BookService;
+
+import java.util.Optional;
+@RequiredArgsConstructor
```
пробел между импортом и аннотацией

RinatBeybutov commented on Jpa/src/main/java/ru/maxima/jpa/controller/AppController.java:
```diff
@@ -0,0 +1,44 @@
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/books")
+public class AppController {
```
давай разнесем на два контроллера методы, которые возвращают название view шаблона и рест методы, которые данные возвращают. В рест методах нужно возвращать не строку, а дто книги, те ResponseEntity<BookDto> .
На вьюхах @Controller, а на рестах - @RestController и напиши ответным комментарием, в чем разница между ними

Sindicat007 replied:
@Controller - используется для обработки входящих запросов, возвращая имя представления(модели, например jsp), которое будет использоваться для отображения ответа. 
Так же можно использовать в сочетании с аннотацией @ResponseBody для обработки запросов, преобразуя их в JSON или XML как в @RestController.

@RestController - используется для обработки входящих запросов и возвращает данные, преобразуя их в JSON или XML.

RinatBeybutov commented on Jpa/src/main/java/ru/maxima/jpa/entity/Book.java:
```diff
@@ -0,0 +1,20 @@
+@Getter
+@Setter
+@Table(schema = "books", name = "book")
+public class Book {
```
давай переименуем в BookEntity

Sindicat007 committed:
- Отформатировал код,
- Переименовал сущность Book в BookEntity,
- Перенес в ViewController методы работающие с view
- Поправил аннотации Маппинга для удаления и добавления новых книг,
- Отформатировал пути в Маппингах

Signed-off-by: Sindicat007 <sindicat007@ya.ru>
